server/src/main/java/org/candlepin/async/tasks/JobCleaner.java

```diff
@@ -18,46 +18,81 @@
 import org.candlepin.async.AsyncJob;
 import org.candlepin.async.JobExecutionContext;
 import org.candlepin.async.JobExecutionException;
-import org.candlepin.model.AsyncJobStatusCurator;
+import org.candlepin.async.JobManager;
+import org.candlepin.common.config.Configuration;
+import org.candlepin.config.ConfigProperties;
+import org.candlepin.model.AsyncJobStatus.JobState;
+import org.candlepin.model.AsyncJobStatusCurator.AsyncJobStatusQueryBuilder;
 import org.candlepin.util.Util;
 
 import com.google.inject.Inject;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.Arrays;
 import java.util.Date;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+
 
 /**
- * JobCleaner removes finished jobs older than yesterday, and failed
- * jobs from 4 days ago.
+ * The JobCleaner job deletes terminal jobs older than the max job age (default: 7 days)
  */
 public class JobCleaner implements AsyncJob {
-
     private static Logger log = LoggerFactory.getLogger(JobCleaner.class);
 
-    private final int MAX_JOB_AGE_IN_DAYS = 4;
-
-    private AsyncJobStatusCurator jobCurator;
     public static final String DEFAULT_SCHEDULE = "0 0 12 * * ?";
+
     public static final String JOB_KEY = "JobCleaner";
     public static final String JOB_NAME = "job cleaner";
 
+    public static final String CFG_MAX_JOB_AGE = "max_job_age_in_minutes";
+    public static final int CFG_DEFAULT_MAX_JOB_AGE = 10080; // 7 days
+
+    private Configuration config;
+    private JobManager jobManager;
+
     @Inject
-    public JobCleaner(AsyncJobStatusCurator jobCurator) {
-        this.jobCurator = jobCurator;
+    public JobCleaner(Configuration config, JobManager jobManager) {
+        this.config = Objects.requireNonNull(config);
+        this.jobManager = Objects.requireNonNull(jobManager);
     }
 
     @Override
     public Object execute(JobExecutionContext context) throws JobExecutionException {
-        Date deadLineDt = Util.yesterday();
-        int oldCompletedJobs = this.jobCurator.cleanUpOldCompletedJobs(deadLineDt);
+        String cfgName = ConfigProperties.jobConfig(JOB_KEY, CFG_MAX_JOB_AGE);
+        int maxAgeInMinutes = this.config.getInt(cfgName, CFG_DEFAULT_MAX_JOB_AGE);
+
+        if (maxAgeInMinutes < 1) {
```

Comment:
Maybe. The example given leaves us in the same state in both cases, since the job class is created right before execution. So what we'd get is a JobInitializationException with the cause of whatever exception we throw from the constructor. In both cases, the system keeps chugging along until someone thinks to check the logs.

I suppose we could do something like add another flag to JobExceptions that let us permanently disable a job if set; but we're still in the same overall condition: the system keeps going until someone looks at the logs. I don't know if the job system has a means of firing an error that will stop CP dead in its tracks.

Reply:
Ah, okay.  I thought that we might end up with Guice shutting everything down because initialization of a bound class failed.